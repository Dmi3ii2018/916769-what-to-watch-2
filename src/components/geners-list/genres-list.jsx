import React from "react";
import PropTypes from "prop-types";
import {setFilter, filterFilms} from "../../actions/actions";
import {connect} from "react-redux";

const INITIAL_GENRE = `All genres`;

const crateFilterList = (state) => {
  const genres = new Set();
  genres.add(INITIAL_GENRE);
  state.filmsList.forEach((it) => genres.add(it.genre));
  console.log(genres);
  console.log(state);

  return Array.from(genres);
};

const mapStateToProps = (state) => ({
  filterName: state.genre,
  filterNamesList: crateFilterList(state),
});

const mapDispatchToProps = (dispatch) => ({
  onGenreClick: (text) => {
    dispatch(setFilter(text));
    dispatch(filterFilms(true));
  }
});

class Genres extends React.PureComponent {
  constructor(props) {
    super(props);
  }

  render() {
    const {filterName, filterNamesList, onGenreClick} = this.props;

    return <ul className="catalog__genres-list">
      {filterNamesList.map((generName) =>
        <li key={generName} className={`catalog__genres-item catalog__genres-item--${generName === filterName ? `active` : ``}`}>
          <a href="#" onClick={(evt) => onGenreClick(evt.target.textContent)} className="catalog__genres-link">{generName}</a>
        </li>
      )
      }
    </ul>;
  }
}

export const GenresList = connect(mapStateToProps, mapDispatchToProps)(Genres);

Genres.propTypes = {
  filterName: PropTypes.string,
  filterNamesList: PropTypes.array,
  filmsList: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number.isRequired,
        genre: PropTypes.string.isRequired,
        name: PropTypes.string.isRequired,
        img: PropTypes.string.isRequired,
        preview: PropTypes.string.isRequired,
      })
  ),
  onGenreClick: PropTypes.func.isRequired,
};
