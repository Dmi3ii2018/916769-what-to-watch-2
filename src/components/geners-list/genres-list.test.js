import React from "react";
import {shallow} from "enzyme";
import {Genres} from "./genres-list";

it(`Filter component renders properly`, () => {
  const tree = shallow(<Genres
    filterName={`Drama`}
    filterNamesList={[`Drama`, `Comedy`, `Crime`]}
    onGenreClick={jest.fn()}
  />);

  expect(tree).toMatchSnapshot();
});

