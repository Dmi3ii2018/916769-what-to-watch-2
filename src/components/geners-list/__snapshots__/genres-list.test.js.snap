// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter component renders properly 1`] = `ShallowWrapper {}`;
